### PR TITLE
【complex op】No.贰〇 add complex support for tanhshrink

### PR DESCRIPTION
**Pull request by Liyulingyue**
### PR types

New features
### PR changes

OPs 
### Description

suppot complex for tanhshrink, without changing in `python/paddle/tensor/ops.py` and `test/legacy_test/test_activation_op.py`

https://github.com/PaddlePaddle/Paddle/issues/56145

### CODE REVIEW

File: paddle/phi/common/complex.h [hunk 1]
```diff
@@ -436,6 +436,16 @@ HOSTDEVICE inline complex<T> sin(const complex<T>& a) {
 #endif
 }
 
+template <typename T>
+HOSTDEVICE inline complex<T> tanhshrink(const complex<T>& a) {
+#if defined(PADDLE_WITH_CUDA_OR_HIP_COMPLEX) && \
+    (defined(__CUDA_ARCH__) || defined(__HIPCC__))
+  return complex<T>(thrust::tanhshrink(thrust::complex<T>(a)));
+#else
+  return complex<T>(std::tanhshrink(std::complex<T>(a)));
```
Comment (ScottWong98) [inline paddle/phi/common/complex.h]: c++ std 库和 thrust 库中应该是没有 tanhshrink 的

File: paddle/phi/common/complex.h [hunk 2]
```diff
@@ -436,6 +436,16 @@ HOSTDEVICE inline complex<T> sin(const complex<T>& a) {
...
+#endif
+}
+
```
Comment (Liyulingyue) [inline paddle/phi/common/complex.h]:
```suggestion
```

File: paddle/phi/kernels/cpu/activation_grad_kernel.cc [hunk 3]
```diff
@@ -300,6 +300,8 @@ PD_REGISTER_ACTIVATION_GRAD_KERNEL(relu6_grad, Relu6GradKernel)
 PD_REGISTER_ACTIVATION_GRAD_KERNEL(softshrink_grad, SoftShrinkGradKernel)
 PD_REGISTER_ACTIVATION_GRAD_KERNEL(hard_shrink_grad, HardShrinkGradKernel)
 PD_REGISTER_ACTIVATION_GRAD_KERNEL(tanh_shrink_grad, TanhShrinkGradKernel)
```
Comment (Liyulingyue) [inline paddle/phi/kernels/cpu/activation_grad_kernel.cc]:
```suggestion
```

File: paddle/phi/kernels/cpu/activation_kernel.cc [hunk 4]
```diff
@@ -194,6 +194,7 @@ PD_REGISTER_ACTIVATION_KERNEL(thresholded_relu, ThresholdedReluKernel)
 PD_REGISTER_ACTIVATION_KERNEL(hard_shrink, HardShrinkKernel)
 PD_REGISTER_ACTIVATION_KERNEL(softshrink, SoftShrinkKernel)
 PD_REGISTER_ACTIVATION_KERNEL(tanh_shrink, TanhShrinkKernel)
```
Comment (Liyulingyue) [inline paddle/phi/kernels/cpu/activation_kernel.cc]:
```suggestion
```

**Review by ScottWong98 — COMMENTED** [review]
Maybe we should add testing cases for complex data type

**Comment by Liyulingyue** [comment]
> Maybe we should add testing cases for complex data type

got it~

File: test/legacy_test/test_activation_op.py [hunk 5]
```diff
@@ -967,6 +967,40 @@ def test_errors(self):
...
 
 
+class TestTanhshrinkComplex(unittest.TestCase):
+    # test paddle.nn.Tanhshrink, paddle.nn.functional.tanhshrink with complex
+    def setUp(self):
+        np.random.seed(1024)
+        self.x_np = np.random.uniform(10, 20, [10, 17]).astype(
+            np.float64
+        ) + 1j * np.random.uniform(10, 20, [10, 17]).astype(np.float64)
+        self.place = (
```
Comment (ScottWong98) [inline test/legacy_test/test_activation_op.py]: `self.place` 在 `TestTanhshrinkComplex` 中应该是没有用到的

File: test/legacy_test/test_activation_op.py [hunk 6]
```diff
@@ -967,6 +967,40 @@ def test_errors(self):
                 F.tanhshrink(x_fp16)
 
 
+class TestTanhshrinkComplex(unittest.TestCase):
```
Comment (ScottWong98) [inline test/legacy_test/test_activation_op.py:980]: 建议添加一个继承 `TestTanhshrink` 类的 Case

File: test/legacy_test/test_activation_op.py [hunk 7]
```diff
@@ -967,6 +967,40 @@ def test_errors(self):
...
+        self.x_np = np.random.uniform(10, 20, [10, 17]).astype(
+            np.float64
+        ) + 1j * np.random.uniform(10, 20, [10, 17]).astype(np.float64)
+        self.place = (
+            paddle.CUDAPlace(0)
+            if paddle.is_compiled_with_cuda()
+            else paddle.CPUPlace()
+        )
+
+    def test_complex64(self):
```
Comment (ScottWong98) [inline test/legacy_test/test_activation_op.py:988]: `def test_complex64` 和 `def test_complex128` 应该都没有在 GPU 上进行测试吧

File: paddle/phi/kernels/funcs/activation_functor.h [hunk 8]
```diff
@@ -1642,6 +1642,24 @@ struct TanhShrinkGradFunctor : public BaseActivationFunctor<T> {
   static constexpr ActBwdOpFwdDeps FwdDeps() { return ActBwdOpFwdDeps::kDepX; }
 };
 
+template <typename T>
+struct TanhShrinkGradFunctor<ComplexType<T>>
```
Comment (ScottWong98) [inline paddle/phi/kernels/funcs/activation_functor.h:1646]: `TanhShrinkGradFunctor` 只针对 CPU，还需要补充 `CudaTanhShrinkGradFunctor` 下的 complex 类型

**Pull request closed without merging**